### PR TITLE
fix: faster stream verification

### PR DESCRIPTION
**Pull request by H4ad**
## Faster integrity check when is stream

I also take a look at streams mode because PNPM also [verify the integrity of the files using streams](https://github.com/pnpm/pnpm/blob/ef6c22e129dc3d76998cee33647b70a66d1f36bf/fetching/tarball-fetcher/src/remoteTarballFetcher.ts#L204-L217).

The initial version was already fast compare to the main:

```
sri.fromStream(stream, largeIntegrity) x 145 ops/sec ±1.58% (78 runs sampled)
ssri.fromStream(stream, tinyIntegrity) x 9,508 ops/sec ±2.53% (76 runs sampled)
ssri.checkStream(stream, largeIntegrity) x 153 ops/sec ±0.92% (78 runs sampled)
ssri.checkStream(stream, tinyIntegrity) x 9,055 ops/sec ±1.81% (80 runs sampled)
```

I also saw that `checkStream` doesn't support the option `single` and almost all verifications that are done by PNPM only verify a single hash, so I see an opportunity to push the performance a little bit further.

```
ssri.fromStream(stream, largeIntegrity) x 147 ops/sec ±1.76% (76 runs sampled)
ssri.fromStream(stream, tinyIntegrity) x 10,339 ops/sec ±2.71% (80 runs sampled)
ssri.checkStream(stream, largeIntegrity) x 152 ops/sec ±1.14% (80 runs sampled)
ssri.checkStream(stream, tinyIntegrity) x 10,023 ops/sec ±1.46% (81 runs sampled)

ssri.checkStream(stream, largeIntegrity, { single: true }) x 151 ops/sec ±1.19% (79 runs sampled)
ssri.checkStream(stream, tinyIntegrity, { single: true }) x 10,278 ops/sec ±1.35% (81 runs sampled)
```

But I did an experiment, If we ignore all the checkStream codes and jump to the final verification, we can achieve this performance:

```
ssri + createHash (largeIntegrity) x 318 ops/sec ±1.52% (81 runs sampled)
ssri + createHash (tinyIntegrity) x 15,863 ops/sec ±1.70% (81 runs sampled)
```

I put the code in the file above, the assumption is: if we verify only one hash, we can skip a lot of verifications.
So I think I could be good to `ssri` to export single hash verifications, what do you think?

<details>
<summary>benchmark-stream.js</summary>

```js
const Benchmark = require('benchmark');
// const wtf = require("wtfnode");
// wtf.init();
const ssri = require('./lib/index');
const suite = new Benchmark.Suite();
const fs = require('fs');
const crypto = require('crypto');
const { Readable } = require('stream');

const largeText = 'a'.repeat(64).repeat(100);
const largeTextSplitted = largeText.split('');

const tinyText = 'a'.repeat(64);
const tinyTextSplitted = tinyText.split('');

const getStream = (text) => Readable.from(text);

function hash(data, algorithm) {
  return crypto.createHash(algorithm).update(data).digest('base64');
}

const largeIntegrity = `sha512-${hash(largeText, 'sha512')}`;
const tinyIntegrity = `sha512-${hash(tinyText, 'sha512')}`;

suite
  .add('ssri.fromStream(stream, largeIntegrity)', {
    defer: true,
    fn: function (deferred) {
      const stream = getStream(largeTextSplitted);

      ssri.fromStream(stream, largeIntegrity).then(() => {
        deferred.resolve();
      });
    },
  })
  .add('ssri.fromStream(stream, tinyIntegrity)', {
    defer: true,
    fn: function (deferred) {
      const stream = getStream(tinyTextSplitted);

      ssri.fromStream(stream, tinyIntegrity).then(() => {
        deferred.resolve();
      });
    },
  })
  .add('ssri.checkStream(stream, largeIntegrity)', {
    defer: true,
    fn: function (deferred) {
      const stream = getStream(largeTextSplitted);

      ssri.checkStream(stream, largeIntegrity).then(() => {
        deferred.resolve();
      });
    },
  })
  .add('ssri.checkStream(stream, tinyIntegrity)', {
    defer: true,
    fn: function (deferred) {
      const stream = getStream(tinyTextSplitted);

      ssri.checkStream(stream, tinyIntegrity).then(() => {
        deferred.resolve();
      });
    },
  })
  .add('ssri.checkStream(stream, largeIntegrity, { single: true })', {
    defer: true,
    fn: function (deferred) {
      const stream = getStream(largeTextSplitted);

      ssri.checkStream(stream, largeIntegrity, { single: true }).then(() => {
        deferred.resolve();
      });
    },
  })
  .add('ssri.checkStream(stream, tinyIntegrity, { single: true })', {
    defer: true,
    fn: function (deferred) {
      const stream = getStream(tinyTextSplitted);

      ssri.checkStream(stream, tinyIntegrity, { single: true }).then(() => {
        deferred.resolve();
      });
    },
  })
  .add('ssri + createHash (largeIntegrity)', {
    defer: true,
    fn: function (deferred) {
      const stream = getStream(largeTextSplitted);
      const parsed = ssri.parse(largeIntegrity, { single: true });
      const hash = crypto.createHash(parsed.algorithm);

      stream.pipe(hash);
      stream.on('end', () => {
        const digest = hash.digest('base64');

        if (parsed.digest !== digest) {
          throw new Error('Integrity check failed');
        }
        deferred.resolve();
      });
    },
  })
  .add('ssri + createHash (tinyIntegrity)', {
    defer: true,
    fn: function (deferred) {
      const stream = getStream(tinyTextSplitted);
      const parsed = ssri.parse(tinyIntegrity, { single: true });
      const hash = crypto.createHash(parsed.algorithm);

      stream.pipe(hash);
      stream.on('end', () => {
        const digest = hash.digest('base64');

        if (parsed.digest !== digest) {
          throw new Error('Integrity check failed');
        }
        deferred.resolve();
      });
    },
  })
  .on('cycle', function (event) {
    console.log(String(event.target));
    // wtf.dump();
  })
  .run({ async: false });
```

</details>

## References

Related to #71

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -29,19 +29,34 @@ class IntegrityStream extends MiniPass {
     this.#getOptions()
 
     // options used for calculating stream.  can't be changed.
-    const algorithms = opts?.algorithms || DEFAULT_ALGORITHMS
-    this.algorithms = Array.from(
-      new Set(algorithms.concat(this.algorithm ? [this.algorithm] : []))
-    )
+    if (opts?.algorithms) {
+      const algorithms = opts.algorithms
```
Comment (wraithgar) [inline lib/index.js:33]: do we need to cast this here if we only use it one time in the next line?

Comment (H4ad) [reply lib/index.js]: Make sense, I will remove this variable.

File: lib/index.js [hunk 2]
```diff
@@ -29,19 +29,33 @@ class IntegrityStream extends MiniPass {
...
-      new Set(algorithms.concat(this.algorithm ? [this.algorithm] : []))
-    )
+    if (opts?.algorithms) {
+      this.algorithms = Array.from(
+        new Set(opts.algorithms.concat(this.algorithm ? [this.algorithm] : []))
+      )
+    } else {
+      this.algorithms = DEFAULT_ALGORITHMS
+
+      if (this.algorithm !== null && this.algorithm !== DEFAULT_ALGORITHMS[0]) {
```
Comment (wraithgar) [inline lib/index.js:39]:
```suggestion
      if (this.algorithm?[0] !== DEFAULT_ALGORITHMS[0]) {
```

File: lib/index.js [hunk 3]
```diff
@@ -29,19 +29,33 @@ class IntegrityStream extends MiniPass {
...
 
   #getOptions () {
     // For verification
     this.sri = this.opts?.integrity ? parse(this.opts?.integrity, this.opts) : null
     this.expectedSize = this.opts?.size
-    this.goodSri = this.sri ? !!Object.keys(this.sri).length : false
-    this.algorithm = this.goodSri ? this.sri.pickAlgorithm(this.opts) : null
+    this.goodSri = this.sri instanceof Integrity
+      ? !!Object.keys(this.sri).length
+      : this.sri instanceof Hash
```
Comment (wraithgar) [inline lib/index.js:53]: What use case is this?

File: lib/index.js (hunk 2 above)
Comment (wraithgar) [reply lib/index.js:39]: Sorry about the edits here I got my wires crossed w/ another line.

File: lib/index.js [hunk 4]
```diff
@@ -159,6 +173,17 @@ class Hash {
     return this.toString()
   }
 
+  match (integrity, opts) {
```
Comment (wraithgar) [inline lib/index.js:176]: This match implementation is bugged.  See #41

I was going to look into this next week when @nlf was available to help.

The underlying issue is that "pickAlgorithm" is perhaps not appropriate here.  If one integrity has sha1 and the other has sha1 and sha512, we simply need to see if the sha1 values match.  pickAlgorithm will only pull out the sha512 from the second integrity so we'll fail the match.

**Comment by wraithgar** [comment]
The nested ternaries in `#getOptions` are extremely hard to follow, can we take this opportunity to remove them?

```javascript
    this.algorithm = null
    if (this.sri instanceof Integrity) {
      this.goodSri = !!Object.keys(this.sri).length
      if (this.goodSri) {
        this.algorithm = this.sri.pickAlgorithm(this.opts)
      }
    } else if (this.sri instanceof Hash) {
      this.goodSri = this.sri
      this.algorithm = this.sri.algorithm
    }
```

File: lib/index.js (hunk 3 above)
Comment (H4ad) [reply lib/index.js:53]: This line: https://github.com/pnpm/pnpm/blob/ef6c22e129dc3d76998cee33647b70a66d1f36bf/fetching/tarball-fetcher/src/remoteTarballFetcher.ts#L206

From what I see from the NPM data, the integrity usually is a 256 hash, and it only is one hash.

So, instead of needing to instantiate an `Integrity` class, we can just instantiate the `Hash` class and save some operations when `single=true`.

But thinking a little further, instead of supporting this behavior, maybe we can create a new API specifically to verify only one hash at a time like I put in the `benchmark` (which gives us almost 2x perf), intended for cases like this.

What do you think? Keep or create a new API?

Comment (wraithgar) [reply lib/index.js]: Ok I'm following you now.  The `single: true` option is already the signal we use to inform ssri that we are only parsing a single hash.  Having `checkStream` not break with this param seems preferable to creating a whole new single hash API.  Once that works in and of itself we can then optimize it as we would have new API.

Does this make sense?  Feel free to advocate for your own opinion here, it is important.

File: lib/index.js [hunk 5]
```diff
@@ -29,19 +29,37 @@ class IntegrityStream extends MiniPass {
...
 
   #getOptions () {
     // For verification
     this.sri = this.opts?.integrity ? parse(this.opts?.integrity, this.opts) : null
     this.expectedSize = this.opts?.size
-    this.goodSri = this.sri ? !!Object.keys(this.sri).length : false
-    this.algorithm = this.goodSri ? this.sri.pickAlgorithm(this.opts) : null
+    this.algorithm = null
+
+    if (this.sri instanceof Integrity) {
```
[the remark this answers is not in the document]
Comment (H4ad) [reply lib/index.js]: The performance is the same so no problem.

File: lib/index.js (hunk 4 above)
Comment (H4ad) [reply lib/index.js:181]: Do you think that worth to hold this PR until the behavior of match is fixed?

Comment (wraithgar) [reply lib/index.js:181]: No I suppose it isn't.  We're not introducing a NEW bug, just carrying it forward with the new behavior.

If you want to help me think through the bug I'd welcome that too.  Basically I need a gut check on my theory that `pickAlgorithm` is not appropriate inside `match`, but instead we need to find if there are any overlaps in hash types between the two comparators, and then make sure the digests match when we find those overlaps.

**Comment by H4ad** [comment]
I was reading the code, check if I'm not wrong, this piece of code:

https://github.com/npm/ssri/blob/8e80eca497bd7f6a714498df6ad64f3faef3bc5e/lib/index.js#L82-L86

And then, the implementation of `match` is:

https://github.com/npm/ssri/blob/8e80eca497bd7f6a714498df6ad64f3faef3bc5e/lib/index.js#L283-L298

What this code is doing:

- Hashing the stream with all the algorithms (from opts or default 512).
- Then, getting the digest and parsing it again to transform to `Integrity` or `Hash`.
- Then, picking just one of the algorithms (basically wasting the time spent to parse it because we just need one).
- And finally, we compare the digest (that we already had)

If the assumptions are:

- We need to compare all the hashes and make sure all match.

So, the code is not implemented correctly.

If the assumptions are:

- We need to compare at least one hash.

So, the code is doing more than needed, we could just parse the sent digest, pick one, and then parse the stream for just that hash.

In the final, I think we could remove entirely the function of `pickAlgorithm` because the assumption behind it is wrong.

File: lib/index.js (hunk 4 above)
Comment (H4ad) [reply lib/index.js:181]: Ok, about the pickAlgorithm, I think I will try to fix it since it could be related to improving the performance of the stream: https://github.com/npm/ssri/pull/76#issuecomment-1496396326

**Comment by wraithgar** [comment]
`pickAlgorithm` is a very very useful function still. Our package fetching library and caching library use it to pull out the "best" algorithm to use when reading fetching/reading data.

Reusing it in `match` I think was where we went wrong.

As far as what `match` *should* be doing, I think I need to think through a boolean chart.

Integrity can have one hash or multiple hashes,  sri can have one hash or multiple hashes.

`Integrity.match(sri)`: 

`(one hash).match(one hash)` - Hash algorithms should match, and digests should match.
`(one hash).match(multiple hashes)` - Integrity hash algorithm should be found in sri hashes, and the digests of that hash should match.
`(multiple hashes).match(one hash)` - sri hash algorithm should be found in Integrity hashes, and the digests of that hash should match.
`(multiple hashes).match(multiple hashes)` - Either we find the "best" algorithm of any matching hashes, and the digests of that one hash should match, or we find all matching algorithms between the Integrity and sri, and all of the digests should match.

**Comment by H4ad** [comment]
@wraithgar Okay, I'll look into this more over the weekend to see if I can get more optimizations out of this.

But those optimizations will be along with the `match` fix, so I guess we could separate it into another PR.

About this PR, the only thing missing is the tests, right? For the bug on `match` we will just forward it to another PR and keep it as far as I can understand.

**Comment by wraithgar** [comment]
Yes by all means let's fix `match` in a separate PR.

**Comment by wraithgar** [comment]
Pinging "Always happy to be pinged" @jakebailey. Not sure if your interests are limited to semver or if they expand to other libs used by pnpm.

Two PRs have already landed here and were published as `v10.0.2`.  I wouldn't bother submitting an update to pnpm quite yet because this PR is likely going to go out sooner than later also.

If you are involved in updating ssri in pnpm please note that `toString()` is no longer returning identical strings as before.  Under the spec this doesn't matter, and as far as I can tell pnpm is not making that string do any heavy lifting on its own, but it's something to be aware of and definitely something we can revisit if it ends up being a problem for pnpm (or yarn).

**Comment by H4ad** [comment]
@jakebailey The PR that changes the behavior of `toString` is this one: https://github.com/npm/ssri/pull/75

Essentially, if the hash was created from a string like `sha256-xxx sha512-xxx`, the output when calling `parsed.toString()` is `sha512-xxx sha-256xxx` only when `strict=true`.

The only time `PNPM` call that method is in this place: https://github.com/search?q=repo%3Apnpm%2Fpnpm%20ssri.parse&type=code

So I think it will not be a problem for PNPM, for other package managers, I don't know if could be a problem because usually the integrity of the packages is just stored using a single algorithm, this little breaking change only affects when we see more than one algorithm with `strict=true`.

**Comment by wraithgar** [comment]
Yes that's correct. As long as folks are using the literal stringified representation but are instead letting ssri parse it they will be fine.

**Comment by jakebailey** [comment]
Sure, I'll defer to you all and @zkochan; my knowledge of this is limited to "I profiled the code and made it faster" 😄

**Comment by wraithgar** [comment]
@H4ad I think linting is all we're waiting on here.  The `.match` bugfix is gonna take @nlf and I some digging to sort out.

File: lib/index.js [hunk 6]
```diff
@@ -159,6 +179,17 @@ class Hash {
     return this.toString()
   }
 
+  match (integrity, opts) {
+    const other = parse(integrity, opts)
+    if (!other) {
+      return false
```
Comment (wraithgar) [inline lib/index.js:185]: looks like we need a test that hits this line.

Comment (H4ad) [reply lib/index.js:185]: This will take more time to add, I will try to add more test tomorrow.

Comment (wraithgar) [reply lib/index.js:185]: No worries!  Really appreciate your contributions lately.

Comment (H4ad) [reply lib/index.js:185]: I push the tests, I also changed a little bit the implementation because I was doing the wrong verification to compare the integrity, tests are always good ahhaaha

Comment (H4ad) [reply lib/index.js:185]: Actually, still have some lines without coverage but I'm not sure how I can properly test those lines

Comment (wraithgar) [reply lib/index.js:185]: I need to tackle the `match` algorithm anyways. I'll fork from this branch to fix it and when it lands it'll land this also.  Test coverage can be our problem.
